### DISCUSSION TITLE
running brew install formula upgrades every installed package that needs updating?

### DISCUSSION THREAD

**Discussion by clhenrick** [Homebrew/discussions#2034] [Everyday usage]
I recently did a `brew update` followed by a `brew install imagemagick`. What followed was that homebrew found that `imagemagick` was already installed so started upgrading it, but then proceeded to upgrade **every single package** that was out of date without asking me to do so. Is this an expected behaviour of homebrew? Have others encountered this behaviour as well? It doesn't seem like a good dev UX to do this without asking to do so first.

**Comment by gromgit** [comment]
> Is this an expected behaviour of homebrew?

Short answer: Yes. https://docs.brew.sh/FAQ#why-does-brew-upgrade-formula-also-upgrade-a-bunch-of-other-stuff

> It doesn't seem like a good dev UX to do this without asking to do so first.

Even as a not-Homebrew-maintainer, I don't agree. Based on too much sysadmin experience, asking for confirmation only works well if the user has the necessary experience to understand both the implications of what's being asked, and what the appropriate response is given their circumstances and desires. Judging from the kind of discussions and issues filed over time, many (perhaps even most) users who see this:
```
WARNING: You are about to upgrade the following 209 formulae:
<REALLY LONG list, because you haven't run `brew upgrade` in a long while>
Proceed? (Y/n)
```
will post yet another issue:
> Um, hey guys, is this normal? What should I do?

And there will be many of these posts because, when given a choice, most folks are afraid of "doing the wrong thing", even when it's really a Hobson's choice (accept it, or you don't get what you want). 

More info & choices doesn't automatically make "good dev UX", however the term is defined.

**Comment by clhenrick** [comment]
I see your point, but in IMO it's still misleading and confusing to call a
command "install" when it really means "first try to install but then
upgrade all installed formula that can be upgraded". Honestly, yes I would
prefer something like "package x is already installed, would you like to
proceed with an upgrade to version y?" and then a message: "n packages can
also be upgraded. Proceed with upgrading?". I can view all packages that
require upgrading if that's what I wish to see via `brew outdated`, and if
I want to filter that list I'm perfectly capable of doing so how I see fit.

Apologies for the use of jargon. A good developer experience (this is often
referred to as "DX" or "Dev UX") refers to developer tooling, frameworks,
etc. being friendly and easy to use (this is one, but not the only, reason
people really like the React JS framework). To me that implies that said
tooling should work as expected, that when I run "install" something
installs; that when I run "upgrade" something upgrades; "remove" –well you
get the picture. If I have a database running on version x and then all of
a sudden I'm upgrading my db software and haven't done a migration and now
I'm unknowingly running the latest version of that db I'm not going to be
happy the next time I access my db...

On the other hand I agree it can be nice to have automated features like
you're suggesting and what I'm complaining about. IMO they should be opt-in
and not pushed onto the user in an obfuscated way like upgrading all
dependencies when attempting to install a single dependency though.

The root of the problem I'm seeing here is that when running "install" on a
single package that is outdated, the "upgrade" command is run, which
defaults to the behavior you linked to (upgrading *all* outdated formulas).
So instead of an "install" like I intended, I now am doing an "upgrade" of
all formula when that wasn't what I wanted to do.

**Reply by SMillerDev** [reply]
> The root of the problem I'm seeing here is that when running "install" on a single package that is outdated, the "upgrade" command is run, which defaults to the behavior you linked to (upgrading *all* outdated formulas). So instead of an "install" like I intended, I now am doing an "upgrade" of all formula when that wasn't what I wanted to do.

This isn't the case though. It only upgrades dependencies of the new software, and any software that also depends on those dependencies. And the more widely used outdated software you have running, the more it'll upgrade.

FYI: If you keep your system up to date (I do a weekly update) you likely won't even get to a state where install upgrades anything.

**Reply by tlk** [reply]
> The root of the problem I'm seeing here is that when running "install" on a single package that is outdated, the "upgrade" command is run

I agree, and I don't think the current behaviour makes sense from a semantic point of view.

I believe that a better UX/DX would be to display a message ala:
_Package "foobar" is already installed. There is an update available. The update can be installed by running `brew upgrade foobar`._

This way there is no need for confirmation or warning messages, and the outcome of the install and upgrade commands are aligned with the semantic meaning of those words.

------

What may add to the confusion is that by default `brew install` will also run `brew update` before attempting to install anything. So if you run `brew install foobar` and come back a year later and run `brew install foobar` again it will pull down updates and upgrade things (and possibly homebrew itself) instead of just telling you that foobar is already installed.

Note that the automagic `brew update` wizardry can be disabled by setting an environment variable: `export HOMEBREW_NO_AUTO_UPDATE=1`. With this set you will be required to run `brew update` to download formula updates.

**Reply by carlocab** [reply]
If you prefer that `brew install foo` does not upgrade `foo` when `foo` is installed but outdated, set `HOMEBREW_NO_INSTALL_UPGRADE`. For example:
```
❯ brew info cargo-instruments
cargo-instruments: stable 0.4.3 (bottled)
[snip]
❯ cargo-instruments --version
cargo-instruments 0.4.2
❯ HOMEBREW_NO_INSTALL_UPGRADE=1 brew install cargo-instruments
Error: cargo-instruments 0.4.2 is already installed
To upgrade to 0.4.3, run:
  brew upgrade cargo-instruments
```

See https://github.com/Homebrew/brew/blob/216eb45a6fefad9dda0a41a8f56fd276c07e60f3/Library/Homebrew/install.rb#L274-L295.

**Reply by SMillerDev** [reply]
It might break things of course, but at least it won't make sure you run support versions of software anymore.

**Reply by tlk** [reply]
> If you prefer that `brew install foo` does not upgrade `foo` when `foo` is installed but outdated, set `HOMEBREW_NO_INSTALL_UPGRADE`

Very neat! Thank you for that tip Carlo!

**Reply by tlk** [reply]
> It might break things

Then run `brew update` and carry on...

**Reply by SMillerDev** [reply]
You'll need to `brew upgrade` too. Often if you install some software it'll need the newest version of it's dependencies, which you won't be getting if you set this variable. So you're opening your self up to a whole lot of issues that'll be very hard to debug.

**Reply by tlk** [reply]
Ah yes, `brew upgrade` being implied by the carry on :-)

Works well for me. Use case being something ala "update and upgrade regularly and often, but only when I ask it to".

**Reply by carlocab** [reply]
I don't think this should break anything, since `brew` does a pretty good job at keeping dependencies in sync, thankfully. As far as I can tell this should be as problematic as neglecting to run `brew upgrade foo` when `foo` is installed but outdated, which isn't too bad.

In particular, `brew install foo` will still upgrade `foo`'s dependencies if `foo` is not installed, even when `HOMEBREW_NO_INSTALL_UPGRADE` is set.

**Comment by carlocab** [comment]
I doubt this actually upgraded literally every installed package that's outdated. What this did is upgraded every installed formula connected to `imagemagick` via the dependency graph. Unfortunately, the component of the dependency graph containing `imagemagick` is large, which is why it's really easy to see `brew` upgrading a lot of packages and assume it must be upgrading everything that's outdated.

To see the list of formulae I talk about above, run
```sh
cat <(brew deps imagemagick) <(brew uses --installed --recursive imagemagick) <(for i in $(brew deps imagemagick); do brew uses --installed --recursive $i; done) | sort | uniq
```

**Comment by clhenrick** [comment]
Thanks everyone, I can confirm that running `install formula` only upgraded
the dependencies of the formula and not every outdated formula homebrew
installed. My mistake. The `HOMEBREW_NO_INSTALL_UPGRADE` flag is useful to
know about as well, appreciate the help.
